pilot/pkg/proxy/envoy/v2/ondemand.go

```diff
@@ -0,0 +1,438 @@
+package v2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	ads "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+)
+
+func newDeltaXdsConnection(peerAddr string, stream DeltaDiscoveryStream) *XdsConnection {
+	return &XdsConnection{
+		pushChannel:  make(chan *XdsEvent),
+		PeerAddr:     peerAddr,
+		Clusters:     []string{},
+		Connect:      time.Now(),
+		deltaStream:  stream,
+		LDSListeners: []*xdsapi.Listener{},
+		RouteConfigs: map[string]*xdsapi.RouteConfiguration{},
+	}
+}
+
+//DeltaDiscoveryStream represents the GRPC connection for any delta streams.
+type DeltaDiscoveryStream interface {
+	Send(*xdsapi.DeltaDiscoveryResponse) error
+	Recv() (*xdsapi.DeltaDiscoveryRequest, error)
+	grpc.ServerStream
+}
+
+// update the node associated with the connection, after receiving a a packet from envoy.
+func (s *DiscoveryServer) initDeltaConnectionNode(discReq *xdsapi.DeltaDiscoveryRequest, con *XdsConnection) error {
+	con.mu.RLock() // may not be needed - once per connection, but locking for consistency.
```

Comment:
This seems quite duplicated with the code in ADS ( except getting some stuff from the delta request ). Any chance we can unify them ? Will be hard to keep in sync.

Reply:
Fixed. Thanks.